Flaky test: TestStartHeartbeat_RespectsContextCancellation races on ticker vs ctx.Done

## Symptom

`TestStartHeartbeat_RespectsContextCancellation` in `harness/internal/core/heartbeat_test.go` intermittently fails on CI with:

```
--- FAIL: TestStartHeartbeat_RespectsContextCancellation (0.06s)
    heartbeat_test.go:100: heartbeat continued after context cancel: 2 at cancel, 3 after
```

Observed on CI run https://github.com/rxbynerd/stirrup/actions/runs/25600486535/job/75153409172 (against `gh-118` at `4348f12`, but the file is unchanged by that PR).

The companion test `TestStartHeartbeat_StopsOnCancel` is structurally identical and is presumably equally fragile under the same conditions; we just haven't observed it flake yet.

## Root cause

The implementation uses the standard cancel-or-tick select:

```go
// harness/internal/core/loop.go:1019
func (l *AgenticLoop) startHeartbeat(ctx context.Context, interval time.Duration) context.CancelFunc {
    ctx, cancel := context.WithCancel(ctx)
    go func() {
        ticker := time.NewTicker(interval)
        defer ticker.Stop()
        for {
            select {
            case <-ctx.Done():
                return
            case <-ticker.C:
                _ = l.Transport.Emit(types.HarnessEvent{Type: "heartbeat"})
            }
        }
    }()
    return cancel
}
```

With a 10 ms interval and the test's 30 ms cancel point, a ticker value can already be pending in `ticker.C` at the moment `cancel()` is invoked. Go's `select` chooses pseudo-randomly between two ready arms, so the goroutine can fire one more `Emit` before exiting on `ctx.Done()`. The test's assertion is `countAfter == countAtCancel` (strict equality), which leaves zero tolerance for this race.

The race is intrinsic to the impl-plus-test pair, not to CI infrastructure. Re-runs work because the random select branch makes the flake non-deterministic.

## Suggested fix (one of, not all)

1. **Drain the ticker on exit** — explicit drain inside the `<-ctx.Done()` arm; doesn't help if a tick was already racing into `Emit` though, so prefer (2) or (3).
2. **Prioritise cancellation** — wrap as a nested `select` with the ctx.Done check first:
   ```go
   select {
   case <-ctx.Done():
       return
   default:
   }
   select {
   case <-ctx.Done():
       return
   case <-ticker.C:
       _ = l.Transport.Emit(types.HarnessEvent{Type: "heartbeat"})
   }
   ```
   Closes the window but still doesn't make the test deterministic if the post-tick branch is already executing.
3. **Loosen the test invariant** — accept up to 1 extra heartbeat after cancel (or assert "no heartbeat after cancel + 2 × interval"). This is the most honest fix because the production guarantee is "stops eventually," not "stops in zero time."

Recommend (3) as the test-side change; (2) is fine as a defence-in-depth but should not be relied on by tests.

## Out of scope

This is unrelated to PR #126 (gh-118 / Azure WIF) — the heartbeat code is unchanged by that branch. Opening as a standalone bug.

--- a/harness/internal/core/heartbeat_test.go
+++ b/harness/internal/core/heartbeat_test.go
@@ -107,3 +107,30 @@ func TestStartHeartbeat_RespectsContextCancellation(t *testing.T) {
 		t.Errorf("heartbeat continued after context cancel beyond tolerance: %d at cancel, %d after (max +1 allowed)", countAtCancel, countAfter)
 	}
 }
+
+func TestStartHeartbeat_CancelBeforeFirstTick(t *testing.T) {
+	rec := &recordingTransport{}
+	loop := &AgenticLoop{
+		Transport: rec,
+	}
+
+	// Cancel the context before startHeartbeat is invoked so the goroutine's
+	// first action is to observe an already-Done context. This exercises the
+	// cold path of the non-blocking pre-check added in ca80ca7: the outer
+	// select's `ctx.Done` arm should fire on the very first iteration, before
+	// any ticker tick is consumed.
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	_ = loop.startHeartbeat(ctx, 10*time.Millisecond)
+
+	// Sleep three ticker intervals: ample time for the goroutine to have
+	// either taken the pre-check exit or, in the worst case, entered the
+	// blocking select before noticing cancellation.
+	time.Sleep(30 * time.Millisecond)
+
+	count := rec.heartbeatCount()
+	if count > 1 {
+		t.Errorf("expected at most 1 heartbeat on cold-path cancel-before-first-tick (pre-check should usually catch ctx.Done with count=0; tolerate 1 to avoid flakiness when the goroutine entered the blocking select before observing cancellation), got %d", count)
+	}
+}

--- a/harness/internal/core/heartbeat_test.go
+++ b/harness/internal/core/heartbeat_test.go
@@ -70,12 +70,15 @@ func TestStartHeartbeat_StopsOnCancel(t *testing.T) {
 
 	countAtStop := rec.heartbeatCount()
 
-	// Wait a bit more and verify no additional heartbeats arrived.
+	// Wait a bit more and verify no additional heartbeats arrived. The
+	// production guarantee is "stops eventually": a ticker tick that has
+	// already been selected before ctx.Done is observed may still fire once
+	// after stop(). Tolerate at most one extra heartbeat post-cancel.
 	time.Sleep(30 * time.Millisecond)
 	countAfter := rec.heartbeatCount()
 
-	if countAfter != countAtStop {
-		t.Errorf("heartbeat continued after cancel: %d at stop, %d after", countAtStop, countAfter)
+	if countAfter-countAtStop > 1 {
+		t.Errorf("heartbeat continued after cancel beyond tolerance: %d at stop, %d after (max +1 allowed)", countAtStop, countAfter)
 	}
 }
 
@@ -93,10 +96,14 @@ func TestStartHeartbeat_RespectsContextCancellation(t *testing.T) {
 
 	countAtCancel := rec.heartbeatCount()
 
+	// Wait a bit more and verify no additional heartbeats arrived beyond
+	// tolerance. The production guarantee is "stops eventually": a ticker
+	// tick that has already been selected before ctx.Done is observed may
+	// still fire once after cancel(). Tolerate at most one extra heartbeat.
 	time.Sleep(30 * time.Millisecond)
 	countAfter := rec.heartbeatCount()
 
-	if countAfter != countAtCancel {
-		t.Errorf("heartbeat continued after context cancel: %d at cancel, %d after", countAtCancel, countAfter)
+	if countAfter-countAtCancel > 1 {
+		t.Errorf("heartbeat continued after context cancel beyond tolerance: %d at cancel, %d after (max +1 allowed)", countAtCancel, countAfter)
 	}
 }

--- a/harness/internal/core/loop.go
+++ b/harness/internal/core/loop.go
@@ -1022,6 +1022,14 @@ func (l *AgenticLoop) startHeartbeat(ctx context.Context, interval time.Duration
 		ticker := time.NewTicker(interval)
 		defer ticker.Stop()
 		for {
+			// Non-blocking pre-check biases the select toward cancellation:
+			// if ctx is already done, exit before racing the ticker. Closes
+			// the common post-cancel-tick window described in issue #128.
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
 			select {
 			case <-ctx.Done():
 				return

--- a/harness/internal/core/loop.go
+++ b/harness/internal/core/loop.go
@@ -1023,7 +1023,7 @@ func (l *AgenticLoop) startHeartbeat(ctx context.Context, interval time.Duration
 		defer ticker.Stop()
 		for {
 			// Non-blocking pre-check biases the select toward cancellation:
-			// if ctx is already done, exit before racing the ticker. Closes
+			// if ctx is already done, exit before racing the ticker. Narrows
 			// the common post-cancel-tick window described in issue #128.
 			select {
 			case <-ctx.Done():

